### PR TITLE
feat: comments custom field [CU-2tn2923]

### PR DESCRIPTION
**Pull request by CodeVoyager**
## Ticket

https://app.clickup.com/t/2tn2923

**NOTE:** This PR is dependent on https://github.com/VirtusLab-Open-Source/strapi-typed/pull/14

## Summary

What does this PR do/solve?

- Custom field for comments

## Test Plan

- install plugin on Strapi version `4.4.0-alpha.0+`
- use content builder or add manually custom field
- create/edit an entity
- custom field should be visible and editable

### CODE REVIEW

**Comment by cyp3rius** [comment]
Task linked: [CU-2tn2923 Strapi 4.4.0 custom field support](https://app.clickup.com/t/2tn2923)

File: admin/src/utils/__tests__/functions.test.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { assertString } from "../functions";
+
+describe("assertString()", () => {
+  it("should allow strings", () => {
+    expect(() => assertString("Content")).not.toThrow();
+  });
+
+  it("should disallow everything that is not a string", () => {
```
Comment (kamilszewczyk0) [inline admin/src/utils/__tests__/functions.test.ts:8]: "reject" rather than "disallow"

File: admin/src/utils/functions.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ValidationError } from "./errors";
+
+export function assertString(value: unknown): asserts value is string {
+  if (typeof value === "string") {
```
Comment (kamilszewczyk0) [inline admin/src/utils/functions.ts:4]: If exclamation mark added, this alone "return" statement is not required

File: admin/src/utils/resolveCommentStatus.ts [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
-import { ToBeFixed } from "../../../types";
 import { COMMENT_STATUS } from "./constants";
 
+type Config = {
+  removed?: boolean;
+  blocked?: boolean;
+  blockedThread?: boolean;
+  approvalStatus?: string;
+  reviewFlowEnabled?: boolean;
+};
```
Comment (kamilszewczyk0) [inline admin/src/utils/resolveCommentStatus.ts:10]: Suggestion: move to strapi-typed

File: server/register/custom-fields/__tests__/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      .toMatchInlineSnapshot(`
+      Object {
+        "name": "comments",
+        "plugin": "comments",
+        "type": "json",
+      }
+    `);
+  });
+  it("should not apply custom field when strapi does not supports it", () => {
+    expect(() => registerCustomFields({ strapi: {} } as any)).not.toThrow();
```
Comment (kamilszewczyk0) [inline server/register/custom-fields/__tests__/index.test.ts:25]: Shouldn't be any alert provided, if custom fields are not suppported?

File: server/register/index.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { StrapiContext } from "strapi-typed";
+import { registerCustomFields } from "./custom-fields";
+
+const register = (context: StrapiContext) => {
```
Comment (kamilszewczyk0) [inline server/register/index.ts:4]: Is this necessary to have here?

File: admin/src/custom-fields/comments/components/input/types.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export type PopulateField = "avatar" | "author";
+
+export type CommentsFieldValue = {
+  commentsNumber: number;
+  renderType?: "FLAT" | "TREE";
+  sortByDate?: "ASC" | "DESC";
+  filterBy?: "DATE_CREATED" | "APPROVAL_STATUS";
+  filterByValue?: string | Date;
+  populate?: Array<PopulateField>;
+};
```
[the remark this answers is not in the document]
Comment (CodeVoyager) [reply admin/src/custom-fields/comments/components/input/types.ts:10]: "strapi-typed" covers functionality of strapi runtime/core. These types are this plugin specific.

File: admin/src/custom-fields/comments/components/input/utils.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import { uniq } from "lodash";
+import {
+  getMessage,
+  APPROVAL_STATUS,
+  ValidationError,
+  assertString,
+} from "../../../../utils";
+import { CommentsFieldValue, PopulateField } from "./types";
+
+export const asString = () => "";
```
[the remark this answers is not in the document]
Comment (CodeVoyager) [reply admin/src/custom-fields/comments/components/input/utils.ts:11]: This is just fill required void with stable reference. From my perspective to satisfy component's quirk.

File: admin/src/custom-fields/comments/components/input/utils.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  }
+  if (state.filterBy && typeof state.filterBy !== "string") {
+    throw new ValidationError("Comments filterBy is not a string");
+  }
+  if (state.filterByValue && typeof state.filterByValue !== "string") {
+    throw new ValidationError("Comments filterByValue is not a string");
+  }
+  if (state.populate && !Array.isArray(state.populate)) {
+    throw new ValidationError("Comments populate is not an array");
+  }
```
[the remark this answers is not in the document]
Comment (CodeVoyager) [reply admin/src/custom-fields/comments/components/input/utils.ts:49]: These fields are not required. But when present, then are validated.

File: admin/src/custom-fields/comments/components/input/utils.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  if (state.sortByDate && typeof state.sortByDate !== "string") {
+    throw new ValidationError("Comments sortByDate is not a string");
+  }
+  if (state.filterBy && typeof state.filterBy !== "string") {
+    throw new ValidationError("Comments filterBy is not a string");
+  }
+  if (state.filterByValue && typeof state.filterByValue !== "string") {
+    throw new ValidationError("Comments filterByValue is not a string");
+  }
```
[the remark this answers is not in the document]
Comment (CodeVoyager) [reply admin/src/custom-fields/comments/components/input/utils.ts:46]: I'll refactor to use forEach or something similar.

File: server/register/custom-fields/__tests__/index.test.ts (hunk 4 above)
Comment (CodeVoyager) [reply server/register/custom-fields/__tests__/index.test.ts:25]: We aim for gradual improvement. Custom file itself is not required to provide base functionality. Will log.warn be enough?

File: admin/src/utils/resolveCommentStatus.ts (hunk 3 above)
Comment (CodeVoyager) [reply admin/src/utils/resolveCommentStatus.ts:10]: Same as above

File: admin/src/custom-fields/comments/components/icon.tsx [hunk 10]
```diff
@@ -0,0 +1 @@
+export const CustomFieldIcon = () => null
```
[the remark this answers is not in the document]
Comment (CodeVoyager) [reply admin/src/custom-fields/comments/components/icon.tsx:1]: It's a placeholder. Target icon will be provided soon.

**Review by cyp3rius — APPROVED** [review]

**Pull request merged**